File: binarysearch.io/November 2020/303_tree_traversal.py
# class Tree:
#     def __init__(self, val, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, root, moves):
        UP = 'UP'
        RIGHT = 'RIGHT'
        LEFT = 'LEFT'
        stack = []
        idx = 0 
        stack.append(root)
        while idx < len(moves):
            if moves[idx] == UP:
                stack.pop()
            elif moves[idx] == RIGHT:
                stack.append(stack[-1].right)
            elif moves[idx] == LEFT:
                stack.append(stack[-1].left)
            idx += 1
        return stack[-1].val 

File: binarysearch.io/303_tree_traversal.cpp
/**
 * class Tree {
 *     public:
 *         int val;
 *         Tree *left;
 *         Tree *right;
 * };
 */
int solve(Tree* root, vector<string>& moves) {
    vector<Tree *> stack;
    for(int i = 0; i < moves.size(); i++){
        if(moves[i] == "RIGHT"){
            stack.push_back(root);
            root = root->right;
        }else if(moves[i] == "LEFT"){
            stack.push_back(root);
            root = root->left;
        }else{
            root = stack[stack.size()-1];
            stack.pop_back();
        }
    }
    return root->val;
}